File: frontend/src/services/practiceDiscussion.ts
// practiceDiscussion.ts — "Discuss with AI": one question's thread, and a message.
//
// CC_TASK_QUESTION_CHAT_v1. Two calls over one address. Every failure THROWS a
// `DiscussionError` carrying the HTTP status, so the dock can tell "the cap is
// reached" (409) from "no reply came back" (503) from "the thread will not load"
// without matching sentences.
//
// ## Checked BY EYE against `backend/src/dto/practice_discussion.rs`
//
// Hand-written types, like every practice service: a field the backend stops
// sending is `undefined` at runtime and `tsc` says nothing, so the parse below
// refuses a payload missing any field the dock renders.

import { API_BASE_URL } from "./api";
import { authFetch } from "./auth";
import { readErrorMessage } from "./fetchUtils";
import { PRACTICE_TIMEOUT_MS } from "./practice";

/**
 * How long a SEND may take: it waits on a model that may think before it writes.
 * The house RAG-class budget (`services/ask.ts` uses the same two minutes).
 */
export const DISCUSS_SEND_TIMEOUT_MS = 120000;

/** One message in the thread, composed by the server. */
export type DiscussionTurn = {
  id: string;
  role: "user" | "model";
  author: string;
  model_id: string | null;
  text: string;
  /** `5:36 pm`, in the case's timezone. */
  when: string;
  /** What a model reply cost; `null` on user turns and when not reported. */
  input_tokens: number | null;
  output_tokens: number | null;
  ms: number | null;
};

/** One model the picker offers. */
export type DiscussModel = {
  model_id: string;
  display_name: string;
  /** `billed` or `local`. */
  billing_class: string;
};

/** Everything the dock needs. */
export type DiscussionPayload = {
  question_id: string;
  scenario_code: string;
  position: number;
  turns: DiscussionTurn[];
  models: DiscussModel[];
  default_model: string;
  max_turns: number;
  model_turns: number;
};

/** A failed discussion call, with the status the dock branches on. */
export class DiscussionError extends Error {
  constructor(
    message: string,
    public readonly status: number,
  ) {
    super(message);
    this.name = "DiscussionError";
  }
}

const path = (questionId: string) =>
  `${API_BASE_URL}/api/practice/questions/${encodeURIComponent(questionId)}/discussion`;

/** Refuse a body missing what the dock renders — a contract breach, not an empty thread. */
function asPayload(body: Partial<DiscussionPayload>): DiscussionPayload {
  if (
    typeof body.question_id !== "string" ||
    typeof body.scenario_code !== "string" ||
    typeof body.position !== "number" ||
    !Array.isArray(body.turns) ||
    !Array.isArray(body.models) ||
    typeof body.default_model !== "string" ||
    typeof body.max_turns !== "number" ||
    typeof body.model_turns !== "number"
  ) {
    throw new DiscussionError(
      "The discussion response is missing fields — backend/frontend contract mismatch.",
      0,
    );
  }
  return body as DiscussionPayload;
}

async function orThrow(response: Response, what: string): Promise<DiscussionPayload> {
  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new DiscussionError(`${what} (HTTP ${response.status}${detail}).`, response.status);
  }
  return asPayload((await response.json()) as Partial<DiscussionPayload>);
}

/** Load one question's thread, its model list and its cap. */
export async function fetchDiscussion(questionId: string): Promise<DiscussionPayload> {
  const response = await authFetch(path(questionId), { timeoutMs: PRACTICE_TIMEOUT_MS });
  return orThrow(response, "The discussion could not be loaded");
}

/** What one message sends. `draft_answer` only when it differs from the saved answer. */
export type DiscussionMessage = {
  text: string;
  model: string;
  draft_answer?: string;
};

/** Send one message; the server stores it, asks the model, and returns the thread. */
export async function sendDiscussion(
  questionId: string,
  message: DiscussionMessage,
): Promise<DiscussionPayload> {
  const response = await authFetch(path(questionId), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(message),
    timeoutMs: DISCUSS_SEND_TIMEOUT_MS,
  });
  return orThrow(response, "The message was not answered");
}

File: backend/src/dto/practice_discussion.rs
//! The wire shape of "Discuss with AI" (CC_TASK_QUESTION_CHAT_v1).
//!
//! Every turn arrives COMPOSED — its author name and its clock time — so the
//! browser formats nothing. The dock's WORDS ride the deck payload the question
//! page already has (`PracticeWordingDto`, `discuss_*`); this carries the thread,
//! the model list and the cap.

use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// One message the dock renders.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DiscussionTurnDto {
    pub id: Uuid,
    /// `user` or `model`.
    pub role: String,
    /// The person's name, or the model's display name.
    pub author: String,
    /// The model that wrote a model turn; `null` on a user turn.
    pub model_id: Option<String>,
    pub text: String,
    /// `5:36 pm`, in the case's timezone.
    pub when: String,
    /// What a model reply cost; `null` on user turns and when not reported.
    /// Surfaced so spend is observable in the dock, not only in SQL.
    pub input_tokens: Option<i32>,
    pub output_tokens: Option<i32>,
    pub ms: Option<i32>,
}

/// One model the picker offers.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DiscussModelDto {
    pub model_id: String,
    pub display_name: String,
    /// `billed` or `local` — which `{cost}` word the footer prints.
    pub billing_class: String,
}

/// Everything the dock needs, in one response (GET and POST alike).
///
/// ## Domain note: the model list rides HERE (GO ruling 2)
///
/// Not `/api/chat/models`, which requires the AI group. The dock requires a login
/// only, so Marie can never be refused a picker she was offered.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DiscussionPayload {
    pub question_id: Uuid,
    /// `S-5` — the dock's subtitle.
    pub scenario_code: String,
    /// The question's place in its scenario's deck, 1-based — `{n}` in the title.
    pub position: u32,
    /// Oldest first.
    pub turns: Vec<DiscussionTurnDto>,
    pub models: Vec<DiscussModelDto>,
    /// The `practice_discuss_default_model` settings row.
    pub default_model: String,
    /// The per-question cap on model replies, and how many are used.
    pub max_turns: u32,
    pub model_turns: u32,
}

/// One message sent to the dock.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DiscussionRequest {
    pub text: String,
    /// A model id from the payload's list; absent = the default row.
    #[serde(default)]
    pub model: Option<String>,
    /// Marie's UNSAVED draft, sent only when it differs from her saved answer.
    /// Given to the model as her current words for this one message; never stored.
    #[serde(default)]
    pub draft_answer: Option<String>,
}
